packages/neos-ui-sagas/src/Publish/index.ts

```diff
@@ -0,0 +1,183 @@
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+import {put, call, select, takeEvery, take, race, all} from 'redux-saga/effects';
+
+import {AnyError} from '@neos-project/neos-ui-error';
+import {NodeContextPath, WorkspaceName} from '@neos-project/neos-ts-interfaces';
+import {actionTypes, actions, selectors} from '@neos-project/neos-ui-redux-store';
+import {GlobalState} from '@neos-project/neos-ui-redux-store/src/System';
+import {FeedbackEnvelope} from '@neos-project/neos-ui-redux-store/src/ServerFeedback';
+import {PublishingMode, PublishingScope} from '@neos-project/neos-ui-redux-store/src/CR/Publishing';
+import {WorkspaceInformation} from '@neos-project/neos-ui-redux-store/src/CR/Workspaces';
+import backend, {Routes} from '@neos-project/neos-ui-backend-connector';
+
+import {makeReloadNodes} from '../CR/NodeOperations/reloadNodes';
+import {updateWorkspaceInfo} from '../CR/Workspaces';
+
+const handleWindowBeforeUnload = (event: BeforeUnloadEvent) => {
+    event.preventDefault();
+    event.returnValue = true;
+    return true;
+};
+
+type PublishingResponse =
+    | {
+        success: {
+            numberOfAffectedChanges: number;
+        }
+    }
+    | { error: AnyError };
+
+export function * watchPublishing({routes}: {routes: Routes}) {
+    const {endpoints} = backend.get();
+    const ENDPOINT_BY_MODE_AND_SCOPE = {
+        [PublishingMode.PUBLISH]: {
+            [PublishingScope.SITE]:
+                endpoints.publishChangesInSite,
+            [PublishingScope.DOCUMENT]:
+                endpoints.publishChangesInDocument
+        },
+        [PublishingMode.DISCARD]: {
+            [PublishingScope.SITE]:
+                endpoints.discardChangesInSite,
+            [PublishingScope.DOCUMENT]:
+                endpoints.discardChangesInDocument
+        }
+    };
+    const SELECTORS_BY_SCOPE = {
+        [PublishingScope.SITE]: {
+            ancestorIdSelector: selectors.CR.Nodes.siteNodeContextPathSelector
+        },
+        [PublishingScope.DOCUMENT]: {
+            ancestorIdSelector: selectors.CR.Nodes.documentNodeContextPathSelector
+        }
+    };
+
+    const reloadAfterDiscard = makeReloadAfterDiscard({routes});
+
+    yield takeEvery(actionTypes.CR.Publishing.STARTED, function * publishingWorkflow(action: ReturnType<typeof actions.CR.Publishing.start>) {
+        const confirmed = yield * waitForConfirmation();
+        if (!confirmed) {
+            return;
+        }
+
+        const {scope, mode} = action.payload;
+        const endpoint = ENDPOINT_BY_MODE_AND_SCOPE[mode][scope];
+        const {ancestorIdSelector} = SELECTORS_BY_SCOPE[scope];
+
+        const workspaceName: WorkspaceName = yield select(selectors.CR.Workspaces.personalWorkspaceNameSelector);
+        const ancestorId: NodeContextPath = yield select(ancestorIdSelector);
+
+        do {
```

Comment:
well i see at least one of us understand redux 🤯 nice work will have to test each scenario as my brain will not be able to run this code in memory